frontend/app/abtests/ABTest.scala

```diff
@@ -0,0 +1,66 @@
+package abtests
+
+import java.time.Duration.ofDays
+
+import abtests.AudienceId.idFor
+import play.api.mvc.{Cookie, Request, RequestHeader}
+
+
+case class Allocation(test: ABTest, variant: String) {
+  lazy val cookie = Cookie(test.cookieName, variant, httpOnly = false)
+}
+
+trait BaseVariant {
+  val slug: String
+}
+
+abstract class ABTest(val slug: String, val audience: AudienceRange, canRun: RequestHeader => Boolean = _ => true) {
+  val CookiePrefix = "gu.membership"
+
+  val abSlug = s"ab.$slug"
+
+  val cookieName = s"$CookiePrefix.$abSlug"
+
+  type Variant <: BaseVariant
+
+  val variants: Seq[Variant]
+
+  lazy val variantsBySlug: Map[String, Variant] = variants.map(v => v.slug -> v).toMap
+
+  def cookieFor(variant: BaseVariant) =
+    Cookie(cookieName, variant.slug, maxAge = Some(ABTest.CookieAge), httpOnly = false)
+
+  def describeParticipation(implicit request: RequestHeader): String =
+    s"ab.$cookieName=${allocate(request).map(_.slug).getOrElse("None")}"
+
+  def allocate(request: RequestHeader): Option[Variant] = for {
+    v <- variantForcedBy(request) orElse variantFromABCookie(request) orElse defaultVariantFor(idFor(request)) if canRun(request)
+  } yield v
+
+  def variantForcedBy(req: RequestHeader): Option[Variant] = req.getQueryString(abSlug).map(abName => variantsBySlug(abName.stripPrefix("ab.")))
```

Comment:
I just preferred the namespacing of chucking `ab.` before all the AB-forcing query parameters, eg:

* https://mem.thegulocal.com/join/supporter/enter-details?ab.merged-registration=merged

rather than:

* https://mem.thegulocal.com/join/supporter/enter-details?merged-registration=merged

...to avoid possible collision with other query parameters.